android: add subtitle supports

yuazhen opened the pull request:
Only timed text substream such as subrip ass and mov_text etc. is supported currently. Picture based subtitle support such as hdmv_pgs_subtitle will be add in the future. External subtitle support will be add in the future too.
The API is like MediaPlayer's. Using mMediaPlayer.setOnTimedTextListener(mOnTimedTextListener); to set a TimedText listener to deal with the subtitle text fired from ff_ffplay. The demo can be found in IjkVideoView.java.

xinzhengzhang commented on android/ijkplayer/ijkplayer-java/src/main/java/tv/danmaku/ijk/media/player/IjkMediaPlayer.java:
```diff
@@ -598,6 +602,8 @@ public int getSelectedTrack(int trackType) {
             case ITrackInfo.MEDIA_TRACK_TYPE_AUDIO:
                 return (int)_getPropertyLong(FFP_PROP_INT64_SELECTED_AUDIO_STREAM, -1);
+            case ITrackInfo.MEDIA_TRACK_TYPE_TIMEDTEXT:
+                return (int)_getPropertyLong(FFP_PROP_INT64_SELECTED_TIMEDTEXT_STREAM, -1);
```
where is getter?

yuazhen replied:
don't understand the question.

xinzhengzhang replied:
I mean FFP_PROP_INT64_SELECTED_TIMEDTEXT_STREAM has not implemented in ff_ffplay.c, and it will always return default value -1

yuazhen replied:
implemented in ff_ffplay.c. maybe ff_ffplay.c was collapsed so you didn't see the file changes.

xinzhengzhang commented on ijkmedia/ijkplayer/android/ijkplayer_jni.c:
```diff
@@ -905,6 +905,12 @@ static void message_loop_n(JNIEnv *env, IjkMediaPlayer *mp)
+        case FFP_MSG_TIMED_TEXT:
+            {
+            jstring text = (*env)->NewStringUTF(env, (char *)msg.obj);
+            J4AC_IjkMediaPlayer__postEventFromNative(env, weak_thiz, MEDIA_TIMED_TEXT, 0, 0, text);
```
add post_event2 is better

yuazhen replied:
ok

yuazhen replied:
add the new commit

xinzhengzhang commented on ijkmedia/ijkplayer/ff_ffmsg_queue.h:
```diff
@@ -132,6 +133,17 @@ inline static void msg_queue_put_simple3(MessageQueue *q, int what, int arg1, in
     msg_queue_put(q, &msg);
 }
 
+inline static void msg_queue_put_simple4(MessageQueue *q, int what, int arg1, int arg2, void *obj)
```
used for what?

xinzhengzhang commented on ijkmedia/ijkplayer/ff_ffplay.h:
```diff
@@ -100,6 +100,7 @@ int       ffp_video_thread(FFPlayer *ffp);
 
 void      ffp_set_video_codec_info(FFPlayer *ffp, const char *module, const char *codec);
 void      ffp_set_audio_codec_info(FFPlayer *ffp, const char *module, const char *codec);
+void      ffp_set_subtitle_codec_info(FFPlayer *ffp, const char *module, const char *codec);
```
where is implement?

yuazhen replied:
implemented in ff_ffplay.c

xinzhengzhang replied:
Didn't find it.
Is there something has not be committed?

yuazhen replied:
maybe ff_ffplay.c was collapsed on this PR so you didn't see the file changes.

xinzhengzhang commented on ijkmedia/ijkplayer/ff_ffplay_def.h:
```diff
@@ -799,6 +787,10 @@ inline static void ffp_notify_msg3(FFPlayer *ffp, int what, int arg1, int arg2)
     msg_queue_put_simple3(&ffp->msg_queue, what, arg1, arg2);
 }
 
+inline static void ffp_notify_msg4(FFPlayer *ffp, int what, int arg1, int arg2, void *obj) {
```
not used

yuazhen replied:
the other 3 notify_msg implementations cannot send objects to java, so I add a new notify_msg to do this. this function is called in ff_ffplay.c to send subtitle text to java.

xinzhengzhang replied:
https://github.com/Bilibili/ijkplayer/pull/2450#discussion_r92795948

xinzhengzhang commented on ijkmedia/ijkplayer/ijkmeta.h:
```diff
@@ -33,11 +33,13 @@
 #define IJKM_KEY_TYPE           "type"
 #define IJKM_VAL_TYPE__VIDEO    "video"
 #define IJKM_VAL_TYPE__AUDIO    "audio"
+#define IJKM_VAL_TYPE__TIMEDTEXT "timedtext"
```
IJKM_VAL_TYPE__SUBTITLE

xinzhengzhang commented on ijkmedia/ijkplayer/ijkmeta.h:
```diff
@@ -33,11 +33,13 @@
 #define IJKM_KEY_BITRATE        "bitrate"
 #define IJKM_KEY_VIDEO_STREAM   "video"
 #define IJKM_KEY_AUDIO_STREAM   "audio"
+#define IJKM_KEY_TIMEDTEXT_STREAM "timedtext"
```
not used

xinzhengzhang commented on ijkmedia/ijkplayer/ff_ffmsg.h:
```diff
@@ -56,6 +57,7 @@
 
 #define FFP_PROP_INT64_SELECTED_VIDEO_STREAM            20001
 #define FFP_PROP_INT64_SELECTED_AUDIO_STREAM            20002
+#define FFP_PROP_INT64_SELECTED_TIMEDTEXT_STREAM        20011
```
subtitle_stream

yuazhen commented:
Subtitle or TimedText: google adds timedtext type in API 16 and subtitle type was added in API 21, and only timedtext can be used currently. Subtitle substream such as mov_text is treated as timedtext. So I think timedtext is more suitable.

xinzhengzhang commented:
As you said in java naming timedtext is more appropriate, but in the ffmpeg this kind of stream has been defined as a subtitle clearly.
Timedtext will result ambiguity in ffplay.

yuazhen commented on ijkmedia/ijkplayer/android/ijkplayer_jni.c:
```diff
@@ -905,6 +905,12 @@ static void message_loop_n(JNIEnv *env, IjkMediaPlayer *mp)
             break;
         case FFP_MSG_PLAYBACK_STATE_CHANGED:
             break;
+        case FFP_MSG_TIMED_TEXT:
```
call ffp_notify_msg4 in video_image_display2 in ff_ffplay.c

xinzhengzhang replied:
sorry, the diff of ff_ffplay.c has been folded and I missed it.

xinzhengzhang commented on ijkmedia/ijkplayer/ff_ffplay.c:
```diff
@@ -637,6 +680,27 @@ static void video_image_display2(FFPlayer *ffp)
+                            else if (sp->sub.rects[0]->ass) {
+                                parse_ass_subtitle(sp->sub.rects[0]->ass, buffered_text);
+                            }
+                            ffp_notify_msg4(ffp, FFP_MSG_TIMED_TEXT, 0, 0, buffered_text);
```
buffered_text is a stack variable
ffp_notify_msg4 will take his life cycle to msg queue.
https://github.com/Bilibili/ijkplayer/pull/2450#discussion_r92795948

yuazhen replied:
I will do some tests on a new callback mechanism to ensure that there're no memory leaks and no life cycle confusion and then post a new commit.

yuazhen commented:
Please review the new commit of new ffp_notify_msg4 implementation.
About subtitle and timedtext, I was consider to keep the same usages with the Android's, so the subtitle type in ffplay was converted into timedtext and passed to the java layer. I don't know if you think this conversion is necessary?

xinzhengzhang commented:
@yuazhen 
I have add an ability for `AVMessage` to control the resource like string.
https://github.com/Bilibili/ijkplayer/commit/e8e88008cf229f90f342bf04115ac75ae3047034
Implement the dealloc callback in `msg_queue_put_simple4 `

About the the naming, I think `timedtext` is ok to me.

yuazhen commented:
Hi @xinzhengzhang 
I added a dealloc callback in msg_queue_put_simple4, please review it.
Also, it's a bit earlier to call msg_free_res in ijkmp_get_msg, so I moved it to message_loop_n in ijkplayer_jni.c.

xinzhengzhang commented on ijkmedia/ijkplayer/ff_ffplay.c:
```diff
@@ -1774,7 +1857,45 @@ static int video_thread(void *arg)
+            frame_queue_push(&is->subpq);
+#ifdef FFP_MERGE
+        } else if (got_subtitle) {
+            avsubtitle_free(&sp->sub);
```
Where to free the &sp->sub ?

yuazhen replied:
see line 463 and 1146 of ff_ffplay.c. it frees subtitle when subtitle has been "rendered".

xinzhengzhang commented:
Applied. It will merge into master after a fully test.
